packages/omgidl-serialization/src/MessageReader.ts

```diff
@@ -270,26 +246,23 @@ export class MessageReader<T = unknown> {
     if (usesMemberHeader) {
       const needsToReadSentinelHeader = !usesDelimiterHeader && usesMemberHeader; // XCDR1 mutable
       // MUTABLE UNION
-      let unionIsEnded = false;
-      if (typeEndOffset != undefined && reader.offset >= typeEndOffset) {
-        // XCDR2
-        unionIsEnded = true;
-      } else if (!options.isCDR2 && this.maybeConsumeSentinelHeader(reader)) {
-        // XCDR1
-        unionIsEnded = true;
+      const atEndOfUnion = typeEndOffset != undefined && reader.offset >= typeEndOffset;
+      // used to check
+      let emHeader: ReturnType<CdrReader["emHeader"]> | undefined = undefined;
+      if (!atEndOfUnion) {
+        // only read emHeader if it's not at the end of the union already
+        emHeader = reader.emHeader();
```

Comment:
is being at the end the only way we know whether there should be an emheader? is there a more direct way to know that doesn't depend on the offset?